cscs-checks/microbenchmarks/latency/src/latency.cpp

```diff
@@ -0,0 +1,230 @@
+#include <pthread.h>
+
+#define PAGESIZE 4096
+#define CACHELINESIZE 64
+
+// only used for runtime estimation
+#define L1 32768
+#define L2 262144
+#define L3 25600*1024
+#define GHZ 3.3
+
+/**********************************************/
+
+struct alignas(CACHELINESIZE) CacheLine
+{
+    CacheLine* payload;
+    char padding[CACHELINESIZE - sizeof(CacheLine*)];
+};
+
+std::vector<unsigned> random_path(unsigned);
+std::vector<unsigned> random_grouped_path(unsigned, unsigned);
+
+void convert_to_pointers(CacheLine*, std::vector<unsigned> const&);
+size_t chainload(CacheLine*, size_t reps);
+size_t estimate_reps(unsigned sz, unsigned l1, unsigned l2, unsigned l3);
+void set_affinity(pthread_t t, int i);
+
+template <class F, class ...Args>
+double time_function(F func, Args&& ...args)
+{
+    auto t1 = std::chrono::high_resolution_clock::now();
+    auto endpoint = func(std::forward<Args>(args)...);
+    auto t2 = std::chrono::high_resolution_clock::now();
+
+    volatile auto result = endpoint;
+    return std::chrono::duration<double>(t2-t1).count();
+}
+
+
+/**********************************************/
+
+int main(int argc, char ** argv)
+{
+    set_affinity(pthread_self(), 0);
+    bool simulate_large_pages = false;
+
+    for (int iarg = 1; iarg < argc; ++iarg)
+    {
+        unsigned buffer_size = std::stoi(argv[iarg]);
+
+        buffer_size = (buffer_size / PAGESIZE + 1) * PAGESIZE;
+
+        unsigned group_size;
+        if (simulate_large_pages) {
+            unsigned npages = buffer_size / PAGESIZE;
+            // use 48 out of 64 entries in TLB-L1
+            unsigned ngroups = npages / 48 + 1;
+            unsigned pages_per_group = npages / ngroups;
+            group_size = pages_per_group * PAGESIZE;
+            buffer_size = group_size * ngroups;
+        }
+        else
+            group_size = buffer_size;
+
+        unsigned num_elements = buffer_size / sizeof(CacheLine);
+        unsigned group_elements = group_size / sizeof(CacheLine);
+
+        size_t reps = estimate_reps(buffer_size, L1, L2, L3);
+        if (argc > 2) reps = std::stoi(argv[2]);
+
+        CacheLine* chain;
+        if (posix_memalign(reinterpret_cast<void**>(&chain),
+                           PAGESIZE, num_elements * sizeof(CacheLine)))
+            throw std::bad_alloc();
+
+        std::vector<unsigned> path = random_grouped_path(num_elements, group_elements);
+        convert_to_pointers(chain, path);
+
+        double test = time_function(chainload, chain, reps/10);
+        reps = reps * 0.25/(test*10);
+
+        double duration = 0;
+        duration += time_function(chainload, chain, reps);
+
+        double ns_per_load = duration / reps * 1e9;
+        double clocks = GHZ * ns_per_load;
+
+        std::cout << "latency (ns): " << ns_per_load
+                  << " clocks: " << clocks << " size " << buffer_size
+                  << " ngroups: " << buffer_size/group_size << std::endl;
+
+        std::free(chain);
+    }
```

Comment:
I would print an error here, e.g., "progname: too few arguments", if the user does not pass any arguments and also print a small usage message.